code/sha2-mb/Hacl.SHA2.Scalar32.fst

```diff
@@ -12,24 +12,178 @@ open Lib.MultiBuffer
+  (acc2: Hacl.Spec.SHA2.Vec.(state_spec a M32)) :
+  Lemma
+  (requires (let open Hacl.Spec.SHA2.Vec in
+     Lib.Sequence.index (state_spec_v acc1) 0 ==
+     Lib.Sequence.index (state_spec_v acc2) 0))
+  (ensures acc1 == acc2) =
+
+  let open Lib.Sequence in
+  let open Lib.IntVector in
+  let open Hacl.Spec.SHA2.Vec in
+  allow_inversion hash_alg;
+  let acc1_s = (state_spec_v acc1).[0] <: lseq (word a) 8 in
+  let acc2_s = (state_spec_v acc2).[0] <: lseq (word a) 8 in
+
+  let aux (i:nat{i < 8}) : Lemma (acc1.[i] == acc2.[i]) =
+    assert (index (vec_v acc1.[i]) 0 == index #(word a) #8 acc1_s i);
+    assert (index (vec_v acc2.[i]) 0 == index #(word a) #8 acc2_s i);
+    assert (index (vec_v acc1.[i]) 0 == index (vec_v acc2.[i]) 0);
+    eq_intro (vec_v acc1.[i]) (vec_v acc2.[i]);
+    vecv_extensionality acc1.[i] acc2.[i] in
+
+  Classical.forall_intro aux;
+  eq_intro acc1 acc2
+
+let hacl_spec_update_224_256 b st: Lemma (ensures
+  Hacl.Spec.SHA2.update SHA2_256 b st ==
+  Hacl.Spec.SHA2.update SHA2_224 b st)
+=
+  calc (==) {
+    Hacl.Spec.SHA2.update SHA2_256 b st;
+  (==) { Hacl.Spec.SHA2.EquivScalar.update_lemma SHA2_256 b st }
+    Spec.Agile.Hash.update SHA2_256 st b;
+  (==) { Spec.SHA2.Lemmas.update_224_256 st b }
+    Spec.Agile.Hash.update SHA2_224 st b;
+  (==) { Hacl.Spec.SHA2.EquivScalar.update_lemma SHA2_224 b st }
+    Hacl.Spec.SHA2.update SHA2_224 b st;
+  }
+
+inline_for_extraction noextract
+val sha224_update: update_vec_t SHA2_224 M32
+
+#push-options "--fuel 0 --ifuel 0 --z3rlimit 300"
+let sha224_update b st =
+  let open Lib.Sequence in
+  let h0 = ST.get () in
+  sha256_update b st;
+  let h1 = ST.get () in
+  begin
+    let st0: lseq (element_t SHA2_256 M32) 8 = as_seq h0 st in
+    let st0_m32 = (state_spec_v st0).[0] <: words_state SHA2_256 in
+    let b0: multiblock_spec SHA2_256 M32 = as_seq_multi h0 b in
+    let st1: lseq (element_t SHA2_256 M32) 8 = as_seq h1 st in
+    let st1_m32 = (state_spec_v st1).[0] <: words_state SHA2_256 in
+    calc (==) {
+      st1_m32;
+    (==) {}
+      (state_spec_v (SpecVec.update #SHA2_256 b0 st0)).[0];
+    (==) { Hacl.Spec.SHA2.Equiv.update_lemma_l #SHA2_256 #M32 b0 st0 0 }
+      Hacl.Spec.SHA2.update SHA2_256 b0.(|0|) st0_m32;
+    (==) { hacl_spec_update_224_256 b0.(|0|) st0_m32 }
+      Hacl.Spec.SHA2.update SHA2_224 b0.(|0|) st0_m32;
+    (==) { Hacl.Spec.SHA2.Equiv.update_lemma_l #SHA2_224 #M32 b0 st0 0 }
+      (state_spec_v #SHA2_224 #M32 (SpecVec.update #SHA2_224 b0 st0)).[0];
+    };
+    state_spec_v_extensionality SHA2_256
+      (SpecVec.update #SHA2_256 b0 st0)
+      (SpecVec.update #SHA2_224 #M32 b0 st0)
+  end
+
+[@CInline]
+val sha224_update_nblocks: update_nblocks_vec_t' SHA2_224 M32
+let sha224_update_nblocks len b st =
+  let open Lib.Sequence in
+  let h0 = ST.get () in
+  sha256_update_nblocks len b st;
+  let h1 = ST.get () in
+  begin
+    let st0: lseq (element_t SHA2_256 M32) 8 = as_seq h0 st in
```

Comment:
Can we move the  `state_spec_v_extensionality`, `hacl_spec_update_224_256`, `hacl_spec_update_nblocks_224_256`, `hacl_spec_update_last_224_256` lemmas along with their `_scalar32` versions to `Hacl.SHA2.Scalar32.Lemmas`? Thanks!

Reply:
Good point, thanks. I moved all the proofs from this module to Hacl.SHA2.Scalar32.Lemmas